rules/group.go

```diff
@@ -601,14 +606,27 @@ func (g *Group) Eval(ctx context.Context, ts time.Time) {
 	g.cleanupStaleSeries(ctx, ts)
 }
 
+func (g *Group) RuleQueryOffset() time.Duration {
+	if g.queryOffset != nil {
+		return *g.queryOffset
+	}
+
+	if g.opts.DefaultRuleQueryOffset != nil {
```

Comment:
Shouldn't we add some pre checks to ensure that we'll be able to `append` that "old" data?
And some pre-check regarding the eval_interval? (delay<eval_interval maybe?)

Reply:
I'm not sure I follow, can you please elaborate?

Reply:
Adding some validation to make sure that the "append in the past" will be possible (ts>minValidTime) otherwise they would be rejected.
Also we may want to prevent setting the delay > `eval_interval`.

Reply:
> I don't quite understand this. Why does it matter? Let's assume we have a 5-minute delay and a 1-minute interval - this seems perfectly fine.

Yes, I'm aware that everything will be shifted. I was just considering that perhaps we might want to "guarantee" that when delay is enabled and Prometheus is shut down, the window of data that won't be evaluated is `<delay+eval_interval<2*eval_interval` (an extra `eval_interval` compared to when delay is disabled). With an unlimited delay, the gap may be larger upon Prometheus shutdown. But if we assume users know what they're doing, we can leave it as is.

> I'm not entirely sure this is necessary; you'd have 3 signals to indicate that something went wrong:

Yes, I just thought that "failing early" would be better, but I think some of the parameters that control the "append in the past" may change dynamically, which could make spotting that more difficult.

Reply:
Your feedback is very valid. Would you mind opening up an issue with your first point? If it becomes a wider problem (I can't think of a world where this makes sense).

I think we can fix this by looking at the last evaluation as we restart, but we've never had this feedback before, which makes me think that this is not as relevant as we think it is and is not a straightforward piece of work.